File: py/idkollen_client/models/ftn.py
from __future__ import annotations

from typing import Annotated, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from idkollen_client.models.common import ApiErrorCode


class _Base(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class FtnAuthRequest(_Base):
    redirect_url: Optional[str] = None
    request_phone: Optional[bool] = None
    request_email: Optional[bool] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class FtnPending(_Base):
    status: Literal["PENDING"]
    id: str
    ref_id: Optional[str] = None
    url: str


class FtnCompleted(_Base):
    status: Literal["COMPLETED"]
    id: str
    ref_id: Optional[str] = None
    ssn: str
    name: str
    given_name: str
    surname: str
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    birth_date: Optional[str] = None
    pid: Optional[str] = None
    bank_id: Optional[str] = None


class FtnFailed(_Base):
    status: Literal["FAILED"]
    id: str
    ref_id: Optional[str] = None
    error: ApiErrorCode


FtnStatus = Annotated[
    FtnPending | FtnCompleted | FtnFailed,
    Field(discriminator="status"),
]

File: csharp/src/Idkollen.Client/Models/Ftn.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Idkollen.Client.Models;

public sealed record FtnAuthRequest
{
    public string? RedirectUrl { get; init; }
    public bool? RequestPhone { get; init; }
    public bool? RequestEmail { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public interface IFtnStatus { }

public sealed record FtnPending(string Id, string? RefId, string Url) : IFtnStatus;

public sealed record FtnCompleted(
    string Id,
    string? RefId,
    string Ssn,
    string Name,
    string GivenName,
    string Surname,
    string? Phone,
    string? Email,
    string? Address,
    string? BirthDate,
    string? Pid,
    string? BankId
) : IFtnStatus;

public sealed record FtnFailed(string Id, string? RefId, string Error) : IFtnStatus;

internal sealed class FtnStatusConverter : JsonConverter<IFtnStatus>
{
    public override IFtnStatus Read(ref Utf8JsonReader reader, Type _, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var status = doc.RootElement.GetProperty("status").GetString();
        var raw = doc.RootElement.GetRawText();
        return status switch
        {
            "PENDING" => JsonSerializer.Deserialize<FtnPending>(raw, options)!,
            "COMPLETED" => JsonSerializer.Deserialize<FtnCompleted>(raw, options)!,
            "FAILED" => JsonSerializer.Deserialize<FtnFailed>(raw, options)!,
            _ => throw new JsonException($"Unknown ftn status: {status}"),
        };
    }

    public override void Write(Utf8JsonWriter w, IFtnStatus v, JsonSerializerOptions o)
        => throw new NotSupportedException();
}
